astropy/table/table.py

```diff
@@ -1069,7 +1069,11 @@ def __array__(self, dtype=None):
         Coercion to a different dtype via np.array(table, dtype) is not
         supported and will raise a ValueError.
         """
-        if dtype is not None:
+        if np.dtype(dtype).kind == 'O':
```

Comment:
More out of curiosity than that it really matters, but isn't `dtype` guaranteed to be a `np.dtype` here? 

Note that you are relying on `dtype(None)` being float -- I think I'd write this as
```
if dtype is not None:
    if np.dtype(dtype) != object:
        raise ...
    out = ...
```
Comparing with `object` because you use that in the creation of `out`.

Not really important but if you want to rebase anyway to run the docs...